### PR TITLE
Improve reject reasons for unmet shielded requirements

### PR DESCRIPTION
**Pull request by str4d**
These reject messages end up bubbling up to users via the RPC interface.
Distinguishing between the various failure cases will help users figure
out why their transaction is being rejected.

Closes zcash/zcash#3114.

### CODE REVIEW

File: src/coins.cpp [hunk 1]
```diff
@@ -914,7 +916,7 @@ CAmount CCoinsViewCache::GetValueIn(const CTransaction& tx) const
     return nResult;
 }
 
-bool CCoinsViewCache::HaveShieldedRequirements(const CTransaction& tx) const
+std::optional<UnsatisfiedShieldedReq> CCoinsViewCache::HaveShieldedRequirements(const CTransaction& tx) const
```
Comment (daira) [inline src/coins.cpp:919]: Is this how we want to represent the equivalent of a `Result` type? It seems potentially error-prone that it boolean-coerces to `false` on success and `true` on failure.

(There was a [proposal to add std::expected](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4015.pdf) but it never went anywhere, so I think we'd want to add our own, either based on that or based on Rust's [std::result](https://doc.rust-lang.org/std/result/). There is also a [similar type in Boost](https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/tutorial/essential/result.html) but I guess we don't want to use that.)

File: src/main.cpp [hunk 2]
```diff
@@ -1542,10 +1564,14 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
...
-        // are the joinsplits' and sapling spends' requirements met in tx(valid anchors/nullifiers)?
-        if (!view.HaveShieldedRequirements(tx))
-            return state.Invalid(error("AcceptToMemoryPool: shielded requirements not met"),
-                                 REJECT_DUPLICATE, "bad-txns-shielded-requirements-not-met");
+        // Are the shielded spends' requirements met?
+        auto unmetShieldedReq = view.HaveShieldedRequirements(tx);
+        if (unmetShieldedReq) {
+            auto rejectCode = ShieldedReqRejectCode(*unmetShieldedReq);
+            auto rejectReason = ShieldedReqRejectReason(*unmetShieldedReq);
+            return state.Invalid(error("AcceptToMemoryPool: shielded requirements not met (%s)", rejectReason),
```
Comment (daira) [inline src/main.cpp:1572]: Consider adding the txid to all error reporting in `AcceptToMemoryPool`:
```suggestion
            return state.Invalid(error("AcceptToMemoryPool: %s shielded requirements not met (%s)", tx.GetHash().ToString(), rejectReason),
```

File: src/main.cpp [hunk 3]
```diff
@@ -2861,10 +2892,14 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
-            // are the JoinSplit's requirements met?
-            if (!view.HaveShieldedRequirements(tx))
-                return state.DoS(100, error("ConnectBlock(): JoinSplit requirements not met"),
-                                 REJECT_INVALID, "bad-txns-joinsplit-requirements-not-met");
+            // Are the shielded spends' requirements met?
+            auto unmetShieldedReq = view.HaveShieldedRequirements(tx);
+            if (unmetShieldedReq) {
+                auto rejectCode = ShieldedReqRejectCode(*unmetShieldedReq);
+                auto rejectReason = ShieldedReqRejectReason(*unmetShieldedReq);
+                return state.DoS(100, error("ConnectBlock(): shielded requirements not met (%s)", rejectReason),
```
Comment (daira) [inline src/main.cpp:2900]: Similarly, consider adding the txid to all error reporting in `CheckTxInputs`.

**Review by daira — APPROVED** [review]
utACK with non-blocking comments.

File: src/coins.cpp (hunk 1 above)
Comment (str4d) [reply src/coins.cpp:919]: I would definitely prefer something more similar to a Rust `Result` for expected errors, and we could add that in a subsequent refactor.

Comment (daira) [reply src/coins.cpp:919]: Filed #4816 .

**Review by daira — APPROVED** [review]
utACK

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
I missed that there was a test which needed updating. CI will return a failure shortly.

**Comment by str4d** [comment]
@zkbot force

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**